### PR TITLE
Add snpeff

### LINKED ISSUE #210: new module: snpeff
I think it would be good to have module for [snpeff](https://bioconda.github.io/recipes/snpeff/README.html)

### PR DESCRIPTION
**Pull request by maxulysse**
## PR checklist

Closes #210 

- [x] This comment contains a description of changes (with reason).
- [x] If you've fixed a bug or added code that should be tested, add tests!
- [x] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [ ] If necessary, include test data in your PR.
- [x] Remove all TODO statements.
- [x] Emit the `<SOFTWARE>.version.txt` file.
- [x] Follow the naming conventions.
- [x] Follow the parameters requirements.
- [x] Follow the input/output options guidelines.
- [x] Add a resource `label`
- [x] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
    - [x] `PROFILE=docker pytest --tag <MODULE> --symlink --keep-workflow-wd`
    - [x] `PROFILE=singularity pytest --tag <MODULE> --symlink --keep-workflow-wd`
    - [x] `PROFILE=conda pytest --tag <MODULE> --symlink --keep-workflow-wd`

### CODE REVIEW

**Review by KevinMenden — COMMENTED** [review]
This looks good and all but what's the deal with all that Docker container and the `build.sh` file? 🤔 
I mean it's on bioconda right?
I probably missed a slack discussion 😁

**Comment by grst** [comment]
I missed that slack discussion, too! Would be great to document this somewhere in the module for future reference.

**Comment by KevinMenden** [comment]
Linking the slack discussion:
https://nfcore.slack.com/archives/CJRH30T6V/p1624619248479500

File: software/snpeff/build.sh [hunk 1]
```diff
@@ -0,0 +1,34 @@
+export GENOME=GRCh37
```
Comment (drpatelh) [inline software/snpeff/build.sh:1]: Can we have a `for` loop here that goes through these entries? Not sure if you can have a list of lists type structure as you would in Python.

File: software/snpeff/environment.yml [hunk 2]
```diff
@@ -0,0 +1,10 @@
+# You can use this file to create a conda environment for this pipeline:
```
Comment (maxulysse) [inline software/snpeff/environment.yml:1]:
```suggestion
# You can use this file to create a conda environment for this module:
```

File: software/snpeff/environment.yml [hunk 3]
```diff
@@ -0,0 +1,10 @@
+# You can use this file to create a conda environment for this module:
+#   conda env create -f environment.yml
+name: nf-core-snpeff-5.0dev
```
Comment (drpatelh) [inline software/snpeff/environment.yml:3]:
```suggestion
name: nf-core-snpeff-5.0
```

File: software/snpeff/main.nf [hunk 4]
```diff
@@ -0,0 +1,57 @@
+// Import generic module functions
+include { initOptions; saveFiles; getSoftwareName } from './functions'
+
+params.options = [:]
+options = initOptions(params.options)
+params.use_cache = false
```
Comment (drpatelh) [inline software/snpeff/main.nf:6]: Same comment as for [VEP](https://github.com/nf-core/modules/pull/547#discussion_r664373643)

File: software/snpeff/main.nf [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+            container "quay.io/biocontainers/snpeff:5.0--hdfd78af_1"
+        }
+    } else {
+        container "nfcore/snpeff:${params.snpeff_tag}"
+    }
+
+    input:
+    tuple val(meta), path(vcf)
+    val (snpeff_db)
+    path (snpeff_cache)
```
Comment (drpatelh) [inline software/snpeff/main.nf:29]:
```suggestion
    tuple val(meta), path(vcf)
    val   db
    path  cache
```

File: software/snpeff/main.nf [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+
+    output:
+    tuple val(meta), path("*.ann.vcf"), emit: vcf
+    path "*.csv",                       emit: report
+    path "*.version.txt",               emit: version
```
Comment (drpatelh) [inline software/snpeff/main.nf:34]:
```suggestion
    tuple val(meta), path("*.ann.vcf"), emit: vcf
    path "*.csv"                      , emit: report
    path "*.version.txt"              , emit: version
```

File: software/snpeff/main.nf [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+    script:
+    def software = getSoftwareName(task.process)
+    def avail_mem = 6
+    if (!task.memory) {
+        log.info '[snpEff] Available memory not known - defaulting to 6GB. Specify process memory requirements to change this.'
+    } else {
+        avail_mem = task.memory.giga
+    }
+    prefix       = options.suffix ? "${meta.id}${options.suffix}" : "${meta.id}"
+    cache        = params.use_cache ? "-dataDir \${PWD}/${snpeff_cache}" : ""
```
Comment (drpatelh) [inline software/snpeff/main.nf:45]:
```suggestion
    def prefix    = options.suffix ? "${meta.id}${options.suffix}" : "${meta.id}"
    cache        = params.use_cache ? "-dataDir \${PWD}/${snpeff_cache}" : ""
```

**Comment by drpatelh** [comment]
Looks great @maxulysse ! Mainly some aesthetic changes for simplicity and to keep things in line with other nf-core/modules :)

File: software/snpeff/meta.yml [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+    - meta:
+        type: map
+        description: |
+            Groovy Map containing sample information
+            e.g. [ id:'test', single_end:false ]
+    - vcf:
+        type: file
+        description: |
+            vcf to annotate
+    - snpeff_db:
```
Comment (drpatelh) [inline software/snpeff/meta.yml:34]:
```suggestion
    - db:
```

File: software/snpeff/meta.yml [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+        type: value
+        description: |
+            which db to annotate with
+    - snpeff_cache:
```
Comment (drpatelh) [inline software/snpeff/meta.yml:38]:
```suggestion
    - cache:
```

File: software/snpeff/meta.yml [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+        type: file
+        description: |
+            path to snpEff cache (optional)
+output:
+    - vcf_ann:
```
Comment (drpatelh) [inline software/snpeff/meta.yml:43]:
```suggestion
    - vcf:
```

**Comment by drpatelh** [comment]
Any idea why all of the tests are still failing?

**Comment by maxulysse** [comment]
> Any idea why all of the tests are still failing?

Checking that out now

File: software/snpeff/main.nf [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+    if (!task.memory) {
+        log.info '[snpEff] Available memory not known - defaulting to 6GB. Specify process memory requirements to change this.'
+    } else {
+        avail_mem = task.memory.giga
+    }
+    def prefix   = options.suffix ? "${meta.id}${options.suffix}" : "${meta.id}"
+    cache        = params.use_cache ? "-dataDir \${PWD}/${snpeff_cache}" : ""
+    """
+    snpEff -Xmx${avail_mem}g \\
+        $snpeff_db \\
```
Comment (maxulysse) [inline software/snpeff/main.nf:48]:
```suggestion
        $db \\
```

**Comment by maxulysse** [comment]
Should be good now.
I removed the md5sum for the csv file as the file contain the date...

**Review by drpatelh — APPROVED** [review]
💥

**Pull request merged**